### PR TITLE
Added CartesianSpeedConstraint

### PR DESCRIPTION
**Pull request by stevegolton**
Added CartesianSpeedConstraint to constrain the magnitude of the Cartesian velocity vectors for a part of the robot.

Changes in this PR:
- Added a new class CartesianSpeedConstraint in constraint/cartesian_constraints.py
- Added example in examples/plot_cartesian_constraints.py

Checklists:
- [ ] Update CHANGELOG.md with a single line describing this PR - I was unable to find this file?

### CODE REVIEW

File: toppra/constraint/cartesian_constraints.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            The dimension of the joint position.
+        """
+        super(CartesianSpeedConstraint, self).__init__(self.invdyn, self.constraintf, self.constraintg, dof)
+        self.fk = fk
+        self.linear_speed_max = linear_speed_max
+        self.angular_speed_max = angular_speed_max
+
+    def invdyn(self, q, dq, ddq):
+        linear_speed, angular_speed = self.fk(q, dq)
+        return np.array([linear_speed**2, angular_speed**2])
```
Comment (hungpham2511) [inline toppra/constraint/cartesian_constraints.py:37]: This seems strange. We shouldn't have to square the speeds nor the speed limits.

**Comment by hungpham2511** [comment]
Hi @stevegolton, many thanks for the PR. I am sorry for not getting back earlier. Was really busy at work in the past weeks. :) 

Could you help to make the CI task call `cpp-python36`? Look like PyKDL was used but not installed.

> Update CHANGELOG.md with a single line describing this PR - I was unable to find this file?

It should be HISTORY.md

File: toppra/constraint/cartesian_constraints.py (hunk 1 above)
Comment (stevegolton) [reply toppra/constraint/cartesian_constraints.py:37]: I thought it was weird too. If I don't square then it doesn't seem to work at all. The below screenshots should both have a speed limits of 1m/s and 1rad/s.

Without squaring

With squaring

**Comment by stevegolton** [comment]
No worries! Likewise, apologies for not getting back to you sooner either. Work was really busy for me too.

I'm struggling a little with the PyKDL dependency issue. I was using this library because I am used to its C++ API, but in hindsight it looks like it isn't well supported and I'm having a bunch of issues getting the python library installed on the `cimg/python:3.6` docker container.

I think I might update the example to use pinocchio instead.

File: toppra/constraint/cartesian_constraints.py (hunk 1 above)
Comment (jmirabel) [reply toppra/constraint/cartesian_constraints.py:37]: I think you have to square it for the reason I mentioned in #101. Following the derivation of #146, you should be able to understand why.

**Comment by hungpham2511** [comment]
I am sure that you don't need to square the cartesian acceleration. Most likely there is a bug somewhere.

Bests,

Hung

________________________________

**Comment by hungpham2511** [comment]
Ah, the constraints are on Cartesian velocity! That's why. The way that you implement it only work if the constraints are second order but velocity constraints are first-order.

Bests,

Hung

________________________________

File: toppra/constraint/cartesian_constraints.py (hunk 1 above)
Comment (stevegolton) [reply toppra/constraint/cartesian_constraints.py:37]: Thanks @jmirabel, I think I've got my head around it now

**Comment by stevegolton** [comment]
Looks like #146  does everything this PR does and more (c++ interface & variable velocity limits) which is great! Thus, there's no need for this PR anymore so I'll close it.

**Comment by mrunaljsarvaiya** [comment]
@stevegolton Did you land up using the cpp-binding to use the cartesian constraints in python or write a pure python constraint using the cpp code as a reference?

**Comment by stevegolton** [comment]
@mrunaljsarvaiya Hey, I ended up using the cpp version natively in the end. I haven't touched the python versions or any of the bindings since this PR.

**Pull request closed without merging**